### ISSUE TITLE
If macro is passed to NB_MODULE(name, module) it is not expanded

### ISSUE THREAD

**Issue by efifogel** [wjakob/nanobind#65]
### Problem description

If 'name' is a macro, and it is passed to NB_MODULE(name, module) it is not expanded, because NB_MODULE is a macro itself.
The solution, if you decide to implement it (and the workaround) is easy, rename the macro NB_MODULE to, say, NB_MODULE_INTERNAL(), introduce a function called NB_MODULE(), and call NB_MODULE_INTERNAL() from NB_MODULE().

I assume that this is the case for other macros, but this one affected me...

In case you are curious, I'm generating bindings for great parts of CGAL (see https://www.cgal.org/). It is a generic library that extensively uses templates. There are many combinations of types that can be instantiated. We support (via CMake setting) the generation of more than one library that consist of bindings, which can co-exist and be imported by the same Python component. The names of these libraries are automatically generated. These names are passed to the code, as macros, to serve as the corresponding module names.

**Comment by wjakob** [comment]
Could this problem not be solved symmetrically in your code by implementing a `NB_MODULE_WRAPPER` that invokes `NB_MODULE`? It was not clear to me why nanobind is the one who needs to change.

**Comment by skoslowski** [comment]
True, this could be implemented the user code. If it were implemented here, one could even pass down the module name from cmake. That is what we

**Comment by skoslowski** [comment]
True, this could be implemented in the user's code. 

What we usually do for our (pybind11) modules is pass the module name from CMake to ensure that the PyInit function name always matches the module output name. Maybe something like that would be appealing for nanobind - a NB_MODULE without the need of specifying the name, if `nanobind_add_module` is used

**Comment by efifogel** [comment]
I don't think my answer adds much, nevertheless here it is.
Yes, it can be implemented in the user code.
However, you can say that about everything...
It is nicer to handle it just once in the library, rather than repeatedly
for every user who needs it.
I do acknowledge that the handling narrows down to 2 lines of code...
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by wjakob** [comment]
I think that this is an example of a feature request that would be accepted in pybind11, but is rejected here. 

This being a "nano" library, feature requests are considered when they unblock some otherwise unfixable state of affairs. Feature creep was a huge problem for me in pybind11, and I want to avoid that at all cost here (even if it's only 2LOC), given that that there is an easy solution in user code.
